aws_health_to_slack/program.py

```diff
@@ -0,0 +1,115 @@
+    slack_channels = _read_google_sheet()
+    events = _aws_health_events()
+    events_by_arn = {event["arn"]: event for event in events}
+
+    for entity in _affected_aws_entities(events):
+        project = entity.get("tags", {}).get("project")
+        if not project:
+            print(f"Error: AWS entity without project tag: {entity}")
+            continue
+
+        channel = slack_channels.get(project)
+        affecting_events = [events_by_arn[arn] for arn in entity["eventArns"]]
+        _post_slack_message(project, channel, entity, affecting_events)
+
+
+@autokitteh.activity
+def _read_google_sheet() -> dict[str, str]:
+    """Read mapping of project tags to Slack channels from Google Sheet."""
+    match = re.match(r"(.*/d/)?([\w-]{20,})", url)
+    if not match:
+        raise ValueError(f"Error: invalid Google Sheet ID in URL {url!r}")
+
+    id = match.group(2)
+    sheets = google_sheets_client("google_sheets_connection").spreadsheets().values()
+    rows = sheets.get(spreadsheetId=id, range="A:B").execute().get("values", [])
+    return {row[0].strip(): row[1].strip() for row in rows}
+
+
+@autokitteh.activity
+def _aws_health_events() -> list[dict]:
+    """List all recent AWS Health events.
+
+    API Reference:
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/health/client/describe_events.html
+    """
+    try:
+        aws = boto3_client("aws_connection", "health")
+        dt = datetime.now(UTC) - timedelta(minutes=1)
+        filter = {"lastUpdatedTime": [{"from": dt}]}
+        # Possible alternative: describe_events_for_organization.
+        resp = aws.describe_events(filter=filter)
+        events = resp.get("events", [])
+
+        nextToken = resp.get("nextToken")
+        while nextToken:
+            resp = aws.describe_events(filter=filter, nextToken=nextToken)
+            events += resp.get("events", [])
+            nextToken = resp.get("nextToken")
+
+        return events
+    except Exception as e:
+        print(f"Error: {e}")
+        return []
+
+
+@autokitteh.activity
+def _affected_aws_entities(events: list[dict]) -> list[dict]:
+    """List all AWS entities affected by the given AWS Health events.
+
+    API Reference:
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/health/client/describe_affected_entities.html
+    """
+    try:
+        aws = boto3_client("aws_connection", "health")
+        arns = [event["arn"] for event in events]
+        # Possible alternative: describe_affected_entities_for_organization.
+        resp = aws.describe_affected_entities(eventArn=arns)
+        entities = resp.get("entities", [])
+
+        nextToken = resp.get("nextToken")
+        while nextToken:
+            resp = aws.describe_affected_entities(eventArn=arns, nextToken=nextToken)
+            entities += resp.get("entities", [])
+            nextToken = resp.get("nextToken")
+
+        return entities
+    except Exception as e:
```

Comment:
not sure catching `Exception` is a good idea. Isn't there anything more specific?

Reply:
I know, catching `Exception` is an antipattern, but I don't have concrete data yet. Added a TODO comment.